### DISCUSSION TITLE
Handler `/var/task/Bref\LaravelBridge\Queue\QueueHandler` doesn't exist

### DISCUSSION THREAD

**Discussion by dnl4** [brefphp/laravel-bridge#111] [General]
The docs example bellow give me an error:

```
constructs:
    jobs:
        type: queue
        worker:
            handler: Bref\LaravelBridge\Queue\QueueHandler
            runtime: php-81
            timeout: 60 # seconds
```

Error:
```
Unknown application error occurred
Runtime.Unknown
```

Then I changed it to 

```
constructs:
    jobs:
        type: queue
        worker:
            handler: Bref\LaravelBridge\Queue\QueueHandler
            runtime: php-81-fpm
            timeout: 60 # seconds
```

Now I get the error:
`Handler `/var/task/Bref\LaravelBridge\Queue\QueueHandler` doesn't exist`

**Comment by mnapoli** [comment]
Can you make sure you are using the latest versions of Bref and the Laravel bridge?

This should be the first version btw, not the second.

**Comment by dnl4** [comment]
I'm using:

```
"bref/bref": "^2.0",
"bref/laravel-bridge": "^2.1",
```

I've just tested again with this code:
```
constructs:
    teste:
        type: queue
        worker:
            handler: Bref\LaravelBridge\Queue\QueueHandler
            runtime: php-81
            timeout: 60 # seconds
```


Then I've fetched the logs with `serverless logs --function testeWorker`:
Here is the full logs:
```
Running "serverless" from node_modules
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
Creating storage directories: /tmp/storage/bootstrap/cache, /tmp/storage/framework/cache, /tmp/storage/framework/views, /tmp/storage/psysh
Running 'php artisan config:cache' to cache the Laravel configuration
Creating storage directories: /tmp/storage/bootstrap/cache, /tmp/storage/framework/cache, /tmp/storage/framework/views, /tmp/storage/psysh
Running 'php artisan config:cache' to cache the Laravel configuration
Creating storage directories: /tmp/storage/bootstrap/cache, /tmp/storage/framework/cache, /tmp/storage/framework/views, /tmp/storage/psysh
Running 'php artisan config:cache' to cache the Laravel configuration
Creating storage directories: /tmp/storage/bootstrap/cache, /tmp/storage/framework/cache, /tmp/storage/framework/views, /tmp/storage/psysh
Running 'php artisan config:cache' to cache the Laravel configuration
Creating storage directories: /tmp/storage/bootstrap/cache, /tmp/storage/framework/cache, /tmp/storage/framework/views, /tmp/storage/psysh
Running 'php artisan config:cache' to cache the Laravel configuration
INFO  Configuration cached successfully.
INFO  Configuration cached successfully.
INFO  Configuration cached successfully.
INFO  Configuration cached successfully.
Bref\LaravelBridge\Queue\QueueHandler
INFO  Configuration cached successfully.
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 328.26 ms Memory Used: 66 MB
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 311.14 ms Memory Used: 66 MB
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 351.30 ms Memory Used: 66 MB
START
Unknown application error occurred
Runtime.Unknown
END Duration: 355.62 ms Memory Used: 66 MB
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 374.54 ms Memory Used: 66 MB
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 690.82 ms Memory Used: 95 MB
START
Unknown application error occurred
Runtime.Unknown
END Duration: 746.28 ms Memory Used: 95 MB
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 711.99 ms Memory Used: 95 MB
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
Bref\LaravelBridge\Queue\QueueHandler
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
START
Unknown application error occurred
Runtime.Unknown
END Duration: 726.31 ms Memory Used: 95 MB
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 846.64 ms Memory Used: 95 MB
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
INIT_START Runtime Version: provided:al2.v17    Runtime Version ARN: arn:aws:lambda:us-east-1::runtime:f35635a04216ba4e6a0e74d3c2db080d443709a8f3f2e71fedf54a4f8756705d
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 337.55 ms Memory Used: 66 MB
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 358.56 ms Memory Used: 66 MB
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 325.59 ms Memory Used: 66 MB
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 392.41 ms Memory Used: 66 MB
Bref\LaravelBridge\Queue\QueueHandler
START
Unknown application error occurred
Runtime.Unknown
END Duration: 321.54 ms Memory Used: 66 MB
```

**Comment by dnl4** [comment]
Does this information help?

```
InvalidArgumentException
Invalid configuration value provided for "token". Expected Aws\Token\TokenInterface|Aws\CacheInterface|array|bool|callable, but got string(884) "IQoJb3J******************"
token: (Aws\Token\TokenInterface|Aws\CacheInterface|array|bool|callable)
Specifies the token used to authorize requests. Provide an
Aws\Token\TokenInterface object, an associative array of "token", and an
optional "expiration" key, `false` to use a null token, or a callable token
provider used to fetch a token or return null. See Aws\Token\TokenProvider
for a list of built-in credentials providers. If no token is provided, the
SDK will attempt to load one from the environment.
at vendor/aws/aws-sdk-php/src/ClientResolver.php:431
```

**Comment by dnl4** [comment]
All the problems above were solved after downgrading aws-sdk-php from `3.263.11` to `3.244.0`

[https://github.com/aws/aws-sdk-php/issues/2567](url)

**Comment by harshal-ringe** [comment]
We are using PHP 7.3 and can't migrate to PHP 8 soon. The compatible versions for PHP 7.3 are `"bref/bref": "1.7.34"`, `"bref/laravel-bridge": "^1.2.3"`. 

As per the documentation, in serverless.yml, we have to provide `handler: Bref\LaravelBridge\Queue\QueueHandler` but QueueHandler.php file is not present inside `vendor/bref/laravel-bridge` folder. So I have provided a `handler: Bref\LaravelBridge\Queue\LaravelSqsHandler` 
But still, I'm getting this error: Handler `/var/task/Bref\LaravelBridge\Queue\LaravelSqsHandler` doesn't exist.

below is my serverless.yml file
service: web-serverless

provider:
    name: aws
    region: us-west-2
    environment:
        APP_ENV: dev

package:
    # Directories to exclude from deployment
    patterns:
        - '!node_modules/**'
        - '!public/storage'
        - '!resources/assets/**'
        - '!storage/**'
        - '!tests/**'

functions:
    # This function runs the Laravel website/API
    web:
        handler: public/index.php
        timeout: 28 # in seconds (API Gateway has a timeout of 29 seconds)
        layers:
            - ${bref:layer.php-73-fpm}
        events:
            - httpApi: '*'
    # This function lets us run artisan commands in Lambda
    artisan:
        handler: artisan
        timeout: 120 # in seconds
        layers:
            - ${bref:layer.php-73} # PHP
            - ${bref:layer.console} # The "console" layer

constructs:
    jobs:
        type: queue
        worker:
            handler: Bref\LaravelBridge\Queue\LaravelSqsHandler
            layers:
              - ${bref:layer.php-73}
            timeout: 60 # seconds

plugins:
    # We need to include the Bref plugin
    - ./vendor/bref/bref
    - serverless-lift

What could be the issue?

**Reply by mnapoli** [reply]
@harshal-ringe you are using an older version of this package, you should follow the old versions of the documentation.

**Reply by harshal-ringe** [reply]
Thanks @mnapoli for your reply, I have migrated the project to the latest PHP and Bref versions and now the error has gone but the exceptions inside the workers are not getting logged on Cloudwatch. If there is any error worker, the job directly fails without logging the actual exception/error in Cloudwatch. It only logs `production.INFO: Job failed Error executing "SendMessage" on "https://sqs.us-west-2.amazonaws.com/serverless-jobs.fifo"`. What changes I would have to make to log exceptions?

**Comment by steverhoades** [comment]
@dnl4 I am seeing the same issue migration from bref v1. to v2. How did downgrading the PHP SDK help here, the page you linked returns a 404. Thanks!

Bref: 2.1.3
Laravel Bridge: 2.1.3

**Reply by steverhoades** [reply]
Here is the error I am receiving. Definitely looks like some wires are being crossed as it's looking for a file location and not a class name?

serverless.yml
```
functions:
  worker:
    handler: Bref\LaravelBridge\Queue\QueueHandler
    ...
```

```
{
    "message": "Handler `/var/task/Bref\\LaravelBridge\\Queue\\QueueHandler` doesn't exist",
    "type": "Bref\\Runtime\\HandlerNotFound",
    "stackTrace": [
        "#0 /var/task/vendor/bref/bref/src/FunctionRuntime/Main.php(26): Bref\\Runtime\\FileHandlerLocator->get('Bref\\\\LaravelBri...')",
        "#1 /opt/bref/bootstrap.php(17): Bref\\FunctionRuntime\\Main::run()",
        "#2 {main}"
    ]
}```

It would appear that the Bref container is falling back to just the FileHandlerLocator, it's unclear at this point why this is happening.

**Reply by steverhoades** [reply]
It would appear that the laravel-bridge `bref-init` isn't being executed thus the container isn't updated on the bref side.

**Reply by mnapoli** [reply]
@steverhoades thanks for investigating, any idea why that file is not loaded? That is weird, Composer should automatically load it 🤔

Could you try running `composer dump-autoload` to see if it changes things?

**Reply by steverhoades** [reply]
@mnapoli I'll look into this again ASAP. I will say that reverting back to my v1 configuration works (explicit handler files).

**Reply by steverhoades** [reply]
Revisiting this a year later and i'll be damned if i'm seeing the same error when trying to upgrade laravel bridge. Was there any further insight on how to debug this issue?

**Reply by steverhoades** [reply]
Disregard, I see the issue when upgrading to v2. The QueueHandler.php file should be used instead of the v1 LaravelSqsHandler. I'm unblocked.

**Comment by Mesuva** [comment]
I'm hitting this issue at the moment and it has completely stumped me.

What's really strange is that I have SQS queues working fine on another project with the same setup, and very similar versions of laravel/bref and laravel-bridge. I've compared these projects, as far as doing diffs on files like serverless-state.json, and the config of the queues really appears the same. 

I can successfully put a job on to the queue, but then when it gets processed I see in the Cloudwatch logs:

`{"message":"Bref\\LaravelBridge\\Queue\\QueueHandler","type":"Illuminate\\Container\\EntryNotFoundException","stackTrace":["#0 \/var\/task\/vendor\/bref\/laravel-bridge\/src\/HandlerResolver.php(46): Illuminate\\Container\\Container->get('Bref\\\\LaravelBri...')","#1 \/var\/task\/vendor\/bref\/bref\/src\/FunctionRuntime\/Main.php(26): Bref\\LaravelBridge\\HandlerResolver->get('Bref\\\\LaravelBri...')","#2 \/opt\/bref\/bootstrap.php(17): Bref\\FunctionRuntime\\Main::run()","#3 {main}"]}`

- So it doesn't get as far as looking at my job, it can't find QueueHandler.
- I've checked /vendor/composer, and can see QueueHandler referenced in autoload_static.php, path looks fine
- I'm on bref 2.1.5, laravel-bridge 2.1.3, and the latest Laravel v10. But I've also tried winding some of those back to the specific locked versions on the working project, with no luck
- I've even tried using the specific version of Laravel itself from the working project.

I'm pretty sure bref-init.php is running, as the logs include before the error `Running 'php artisan config:cache' to cache the Laravel configuration`, and that appears to be an output from that script. We then get a  `INFO  Configuration cached successfully.`

- The error message refers in the stack trace to Bref\LaravelBridge\HandlerResolver, so it appears that LaravelBridge is being picked up and has been autoloaded.  (I think that's set up in bref-init.php too)

- My queue is configured in serverless.yml as per the docs:
```
constructs:
  jobs:
    type: queue
    worker:
      handler: Bref\LaravelBridge\Queue\QueueHandler
      runtime: php-81
      timeout: 60 # seconds
```

- in my deployment script, I'm using `composer install --optimize-autoloader --no-dev`
- I'm running PHP8.1

The rest of the application in general is working fantastically in production. I'm doing this work on a staging copy, but the setup is pretty much identical (and the staging application works fine too, it's just the queue handling).

I can't think what else to check/upgrade/compare.
@mnapoli - I'd be very interested to pay for further enterprise support, even if it's just for more ideas of what to investigate!

**Reply by mnapoli** [reply]
Hey Ryan!

That's interesting because the error message is different from above. As you identified, the Laravel bridge seems to be correctly loaded, and `bref-init.php` is running.

I tried a few things, just to be sure:

- redeployed a Laravel example with the basic queue config you are using: it's working
- updated everything to latest version, retried: it's still working
- tried running the `composer install --optimize-autoloader --no-dev` command before redeploying: still working

Looking at the error message, I had a doubt that maybe the class name is "double-escaped" and had extra backslashes: `Bref\\LaravelBridge\\Queue\\QueueHandler`. I tried adding quotes in `serverless.yml`, but was not able to deploy something broken (i.e. it was always working).

So to see if that could be it, I forced my service to deploy with `handler: Bref\\LaravelBridge\\Queue\\QueueHandler` and check the error message:

<img width="1364" alt="Screen-001409" src="https://github.com/brefphp/laravel-bridge/assets/720328/7835cb68-44e2-4a31-a7fd-8684084af0f1">

As we can see, the error message is not the same as yours, and clearly contains extra `\\`. So I don't think that's the problem.

I'm running out of ideas, what I would do is check in the AWS Lambda console if the `Handler` config is actually `Bref\LaravelBridge\Queue\QueueHandler` (no weird thing in there). I'd also check in the interactive code editor (if it loads) that the code uploaded looks correct.

Also worth checking: are you deploying any extra `bootstrap/cache` file maybe? Here's what it looks like for me:

<img width="249" alt="Screen-001410" src="https://github.com/brefphp/laravel-bridge/assets/720328/b9acb726-8135-43df-8ea6-6778bac8b8fa">

Can you also confirm you are using only 1 queue?

Are you on Windows, or anything that could lead to something weird? Are you using the latest version of the `serverless` CLI? (I'm on 3.34.0). Which Node version (I'm on v20.2.0)? (I'm not sure how that could relate… but 🤷)

If we can't figure it out here we could discuss looking into your codebase + AWS account, but maybe there's a simple explanation we can pinpoint first.

**Reply by mnapoli** [reply]
Also are you deploying a Docker image?

And I played with tinker just to see if the class was loading. I did that locally, I'm getting an error (not sure why) but do you get the same thing?

<img width="1900" alt="Screen-001411" src="https://github.com/brefphp/laravel-bridge/assets/720328/005c9ac8-b60d-4c37-9c86-5466d021cc9a">

**Reply by Mesuva** [reply]
Howdy Matthieu, I _really_ appreciate your thoughts here!

- I thought about the double escaping too, and did try putting quotes around the handler, but no luck there. I've even thought about weird unicode characters from a cut and paste.
- I had checked the handler in Lambda, looks fine. Code is too big to edit, but overall the application works fine. If something else was corrupt more broadly with composer I reckon the app would be breaking in other places.
![Screen Shot 2023-11-02 at 1 50 17 pm](https://github.com/brefphp/laravel-bridge/assets/1079600/4a5d0420-5233-4c48-9ce5-85848557a731)
- not deploying via Docker, but I did try just in normal controller doing something like:
```
   $container = app();
   $handler = $container->get( 'Bref\LaravelBridge\Queue\QueueHandler' );
```
That throws an error, whilst I can target other classes within Laravel Bridge without error.. (but I admit I'm not sure if it's supposed to work or not)
- I am already running on this project a database based queue, but for testing I've just changed QUEUE_CONNECTION to 'ses', but only one SES queue as such. My goal is to run both, sending some jobs to the database one, and some to this new SES one. 
- Running on mac, and deploying 6 other Laravel/bref apps serverless from my machine. Nothing unusual I can think of!
- I've updated composer, but not thought to check serverless itself. Will check that.  
- I do have two more files in /bootstrap/cache, which are _livewire-components.php_ and _events.php_. But I have both of those in my working project as well.
- I'm planning to try something similar by adding a queue to another project I'm working on

The only 'unusual' thing about this project I guess is that it's one I've worked on over quite some time, maybe a year and a half, and updated along the way. There must be some little override or config somewhere I've missed. 

Going to test this process on my other project next.

**Reply by Mesuva** [reply]
Just updated to serverless 3.36.0, no change.

Also just tried a SQS queue on a new project, worked as expected!

**Reply by mnapoli** [reply]
Oh, that's good at least.

Here is what the Laravel-Bref bridge is doing:

```php
        $this->app->when(QueueHandler::class)
            ->needs('$connection')
            ->giveConfig('queue.default');
```

Maybe try adding that into your own ServiceProvider? That's not ideal, but maybe it will be enough?

Maybe the project is old enough that some automatic registration (service provider? service class?) that has been added in Laravel since isn't happening.

Also I would indeed expect this to work in a controller, so there's definitely something wrong:

```php
   $container = app();
   $handler = $container->get( 'Bref\LaravelBridge\Queue\QueueHandler' );
```

**Reply by Mesuva** [reply]
Well, I've got it working.

And it's probably the most _frustrating_ mistake I've made in the last year. It's almost comical.

What made it hard to pick up was that the EntryNotFoundException exception is very misleading, it's more of a fall back exception, than something that's true.

What I did was edit /laravel/framework/src/Illuminate/Container/Container.php, and adjust the `get()` function there, to just throw $e as soon as it was caught. Otherwise the specific exception is ignored, and you get a EntryNotFoundException.

When I did that, I got:
 
`Fatal error: InvalidArgumentException: The [ses] queue connection has not been configured. in /var/task/vendor/laravel/framework/src/Illuminate/Queue/QueueManager.php:159`

Spot the issue?

In my severless.yml file, I accidentally set QUEUE_CONNECTION to 'ses', and not 'sqs'.
I didn't spot it, as I set some variables like this:
`QUEUE_CONNECTION: ${self:custom.environments.QUEUE_CONNECTION.${sls:stage}}`

So ONE letter mistake, has cost me a day or so's worth of debugging. 
At least I dug through the code a fair bit and have a better understanding of how it all works!
What a doofus.

Thanks again for the suggestions for debugging, I owe you another 🍺 @mnapoli !

The only issue now I'm hitting is that I doesn't look like I can set the default to the database connection, but send to the SQS queue. It looks like you _have_ to set the SQS as the default for this to work. I may have to try it the other way around, where I specifically send jobs to the database queue.  (or I rework that and just use SQS)

**Reply by LewisW** [reply]
@Mesuva Don't worry, you are not the only careless developer out there.... I had made the same mistake and you saved me a lot of pain! Thank you.

**Comment by moynzzz** [comment]
I had the same error message for another reason. So the problem in my case was my dockerfile was wrongly extending the `bref/php-82-console` image instead of `bref/php-82`. So the console handler is to execute artisan, the FPM handler is for website so for the function handlers you need to use `bref/php-*` docker image or the lambda layer.

Hope this will save some time to somebody else.

**Reply by michaelaguiar** [reply]
Would you mind sharing an example of your queue Dockerfile? I am trying to figure out how to get the QueueHandler to work in the container!

**Reply by moynzzz** [reply]
My configuration is the same as in the docs explained [here](https://bref.sh/docs/deploy/docker). My dockerfile looks like this:

```docker
FROM --platform=linux/amd64 bref/php-82:2

# Some php extensions (ignore this if you dont need php extensions)
COPY --from=bref/extra-gd-php-82:1 /opt /opt
COPY --from=bref/extra-gmp-php-82:1 /opt /opt

COPY . /var/task

CMD ["Bref\\LaravelBridge\\Queue\\QueueHandler"]
```

**Reply by michaelaguiar** [reply]
Thanks for the info! So now I am getting the same error discussed in this thread: 

```json
"message": "Bref\\LaravelBridge\\Queue\\QueueHandler",
"type": "Illuminate\\Container\\EntryNotFoundException",
``` 

Here is my Dockerfile:
```dockerfile
FROM bref/php-83:2

COPY --from=bref/extra-pgsql-php-83:1 /opt /opt
COPY . /var/task

RUN echo "extension=pdo_pgsql.so" >> /opt/bref/etc/php/conf.d/bref.ini

CMD ["Bref\\LaravelBridge\\Queue\\QueueHandler"]

``` 

You have any idea what would cause this? Appreciate it!

**Reply by michaelaguiar** [reply]
Never mind, it's working now. Thanks!

**Reply by zhenghaoweii** [reply]
@michaelaguiar do you mind to share how you able to make it work?

**Reply by michaelaguiar** [reply]
Hey @zhenghaoweii, sorry just noticed this. I was able to get it to work with the Dockerfile above. Are you running into issues still?
